File: modules/platforms/dotnet/Apache.Ignite.Core.Tests/Client/Cache/CreateCacheTest.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache.Ignite.Core.Tests.Client.Cache
{
    using System;
    using System.Linq;
    using System.Threading;
    using Apache.Ignite.Core.Cache.Configuration;
    using Apache.Ignite.Core.Cache.Expiry;
    using Apache.Ignite.Core.Client;
    using Apache.Ignite.Core.Client.Cache;
    using Apache.Ignite.Core.Configuration;
    using Apache.Ignite.Core.Impl.Cache.Expiry;
    using Apache.Ignite.Core.Impl.Client.Cache;
    using Apache.Ignite.Core.Impl.Client;
    using Apache.Ignite.Core.Tests.Cache;
    using NUnit.Framework;

    /// <summary>
    /// Tests dynamic cache start from client nodes.
    /// </summary>
    public class CreateCacheTest : ClientTestBase
    {
        /** Template cache name. */
        private const string TemplateCacheName = "template-cache-*";

        /// <summary>
        /// Tears down the test.
        /// </summary>
        [TearDown]
        public void TearDown()
        {
            DestroyCaches();
        }

        /// <summary>
        /// Destroys caches.
        /// </summary>
        private void DestroyCaches()
        {
            foreach (var cacheName in Client.GetCacheNames())
            {
                Client.DestroyCache(cacheName);
            }
        }

        /// <summary>
        /// Tests the GetCacheNames.
        /// </summary>
        [Test]
        public void TestGetCacheNames()
        {
            DestroyCaches();
            Assert.AreEqual(0, Client.GetCacheNames().Count);

            Client.CreateCache<int, int>("a");
            Assert.AreEqual("a", Client.GetCacheNames().Single());

            Client.CreateCache<int, int>("b");
            Assert.AreEqual(new[] {"a", "b"}, Client.GetCacheNames().OrderBy(x => x).ToArray());

            Client.DestroyCache("a");
            Assert.AreEqual("b", Client.GetCacheNames().Single());
        }

        /// <summary>
        /// Tests create from template.
        /// </summary>
        [Test]
        public void TestCreateFromTemplate()
        {
            // No template: default configuration.
            var cache = Client.CreateCache<int, int>("foobar");
            AssertExtensions.ReflectionEqual(new CacheClientConfiguration("foobar"), cache.GetConfiguration());

            // Create when exists.
            var ex = Assert.Throws<IgniteClientException>(() => Client.CreateCache<int, int>(cache.Name));
            Assert.AreEqual(
                "Failed to start cache (a cache with the same name is already started): foobar", ex.Message);
            Assert.AreEqual(ClientStatusCode.CacheExists, ex.StatusCode);

            // Template: custom configuration.
            cache = Client.CreateCache<int, int>(TemplateCacheName.Replace("*", "1"));
            var cfg = cache.GetConfiguration();
            Assert.AreEqual(CacheAtomicityMode.Transactional, cfg.AtomicityMode);
            Assert.AreEqual(3, cfg.Backups);
            Assert.AreEqual(CacheMode.Partitioned, cfg.CacheMode);
        }

        /// <summary>
        /// Tests getOrCreate from template.
        /// </summary>
        [Test]
        public void TestGetOrCreateFromTemplate()
        {
            // No template: default configuration.
            var cache = Client.GetOrCreateCache<int, int>("foobar");
            AssertExtensions.ReflectionEqual(new CacheClientConfiguration { Name = "foobar"}, cache.GetConfiguration());
            cache[1] = 1;

            // Create when exists.
            cache = Client.GetOrCreateCache<int, int>("foobar");
            Assert.AreEqual(1, cache[1]);

            // Template: custom configuration.
            cache = Client.GetOrCreateCache<int, int>(TemplateCacheName.Replace("*", "1"));
            var cfg = cache.GetConfiguration();
            Assert.AreEqual(CacheAtomicityMode.Transactional, cfg.AtomicityMode);
            Assert.AreEqual(3, cfg.Backups);
            Assert.AreEqual(CacheMode.Partitioned, cfg.CacheMode);

            // Create when exists.
            cache[1] = 1;
            cache = Client.GetOrCreateCache<int, int>(cache.Name);
            Assert.AreEqual(1, cache[1]);
        }

        /// <summary>
        /// Tests cache creation from configuration.
        /// </summary>
        [Test]
        public void TestCreateFromConfiguration()
        {
            // Default config.
            var cfg = new CacheClientConfiguration("a");
            var cache = Client.CreateCache<int, int>(cfg);
            AssertExtensions.ReflectionEqual(cfg, cache.GetConfiguration());

            // Create when exists.
            var ex = Assert.Throws<IgniteClientException>(() => Client.CreateCache<int, int>(cfg));
            Assert.AreEqual(
                "Failed to start cache (a cache with the same name is already started): a", ex.Message);
            Assert.AreEqual(ClientStatusCode.CacheExists, ex.StatusCode);

            // Custom config.
            cfg = GetFullCacheConfiguration("b");

            cache = Client.CreateCache<int, int>(cfg);
            AssertClientConfigsAreEqual(cfg, cache.GetConfiguration());
        }

        /// <summary>
        /// Test cache creation from configuration with expiry policy.
        /// </summary>
        [Test]
        public void TestCreateFromConfigurationWithExpiration()
        {
            var expiryPolicy = new ExpiryPolicy(
                TimeSpan.FromMilliseconds(20),
                TimeSpan.FromMilliseconds(10),
                TimeSpan.FromMilliseconds(30));

            // Default config.
            var cfg = new CacheClientConfiguration("a")
            {
                ExpiryPolicyFactory = new ExpiryPolicyFactory(expiryPolicy)
            };
            var cache = Client.CreateCache<int, int>(cfg);
            AssertExtensions.ReflectionEqual(cfg, cache.GetConfiguration());

            var remoteExpiryPolicy  = cache.GetConfiguration().ExpiryPolicyFactory.CreateInstance();
            AssertExtensions.ReflectionEqual(expiryPolicy, remoteExpiryPolicy);

            cache.Put(1, 1);

            // Wait for expiration period.
            Thread.Sleep(100);

            Assert.IsFalse(cache.ContainsKey(1));
        }

        /// <summary>
        /// Tests cache creation from partial configuration.
        /// </summary>
        [Test]
        public void TestCreateFromPartialConfiguration()
        {
            // Default config.
            var cfg = new CacheClientConfiguration("a") {Backups = 7};
            var client = (IgniteClient) Client;

            // Create cache directly through a socket with only some config properties provided.
            client.Socket.DoOutInOp<object>(ClientOp.CacheCreateWithConfiguration, ctx =>
            {
                var w = ctx.Writer;

                w.WriteInt(2 + 2 + 6 + 2 + 4);  // config length in bytes.

                w.WriteShort(2);  // 2 properties.
                
                w.WriteShort(3);  // backups opcode.
                w.WriteInt(cfg.Backups);

                w.WriteShort(0);  // name opcode.
                w.WriteString(cfg.Name);

            }, null);
            
            var cache = new CacheClient<int, int>(client, cfg.Name);

            AssertExtensions.ReflectionEqual(cfg, cache.GetConfiguration());
        }

        /// <summary>
        /// Tests cache creation from configuration.
        /// </summary>
        [Test]
        public void TestGetOrCreateFromConfiguration()
        {
            // Default configur.
            var cfg = new CacheClientConfiguration("a");
            var cache = Client.GetOrCreateCache<int, int>(cfg);
            AssertExtensions.ReflectionEqual(cfg, cache.GetConfiguration());
            cache[1] = 1;

            // Create when exists.
            cache = Client.GetOrCreateCache<int, int>("a");
            Assert.AreEqual(1, cache[1]);

            // Custom config.
            cfg = GetFullCacheConfiguration("b");

            cache = Client.GetOrCreateCache<int, int>(cfg);
            AssertClientConfigsAreEqual(cfg, cache.GetConfiguration());
        }

        /// <summary>
        /// Gets the full cache configuration.
        /// </summary>
        private static CacheClientConfiguration GetFullCacheConfiguration(string name)
        {
            return new CacheClientConfiguration(CacheConfigurationTest.GetCustomCacheConfiguration(name), true);
        }

        /** <inheritdoc /> */
        protected override IgniteConfiguration GetIgniteConfiguration()
        {
            return new IgniteConfiguration(base.GetIgniteConfiguration())
            {
                CacheConfiguration = new[]
                {
                    new CacheConfiguration(TemplateCacheName)
                    {
                        AtomicityMode = CacheAtomicityMode.Transactional,
                        Backups = 3
                    }
                },
                DataStorageConfiguration = new DataStorageConfiguration
                {
                    DataRegionConfigurations = new[]
                    {
                        new DataRegionConfiguration
                        {
                            Name = "myMemPolicy"
                        } 
                    }
                }
            };
        }
    }
}

File: modules/gatling-plugin/src/test/scala/org/apache/ignite/gatling/CreateCacheTest.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.ignite.gatling

import com.typesafe.scalalogging.StrictLogging
import io.gatling.core.Predef._
import org.apache.ignite.client.ClientCacheConfiguration
import org.apache.ignite.configuration.CacheConfiguration
import org.apache.ignite.gatling.Predef._
import org.apache.ignite.gatling.utils.AbstractGatlingTest
import org.apache.ignite.gatling.utils.IgniteClientApi.NodeApi
import org.apache.ignite.gatling.utils.IgniteClientApi.ThinClient
import org.apache.ignite.gatling.utils.IgniteSupport
import org.junit.Test

/**
 * Tests various create cache operations.
 */
class CreateCacheTest extends AbstractGatlingTest {
  /** Create cache via thin client and simple DSL parameters. */
  @Test
  def thinClientDSL(): Unit = runWith(ThinClient)("org.apache.ignite.gatling.CreateCacheDSLSimulation")

  /** Create cache via thick client and simple DSL parameters. */
  @Test
  def thickClientDSL(): Unit = runWith(NodeApi)("org.apache.ignite.gatling.CreateCacheDSLSimulation")

  /** Create cache via thin client and full configuration instance. */
  @Test
  def thinClientConfig(): Unit = runWith(ThinClient)("org.apache.ignite.gatling.CreateCacheThinConfigSimulation")

  /** Create cache via thick client and full configuration instance. */
  @Test
  def thickClientConfig(): Unit = runWith(NodeApi)("org.apache.ignite.gatling.CreateCacheThickConfigSimulation")
}

/**
 * Create cache with simple parameters specified via DSL.
 */
class CreateCacheDSLSimulation extends Simulation with IgniteSupport with StrictLogging {
  private val cache = "cache"
  private val scn = scenario("Basic")
    .feed(feeder)
    .ignite(
      create(s"$cache-1") as "1",
      create(s"$cache-2").backups(1) as "2",
      create(s"$cache-3").atomicity(TRANSACTIONAL) as "3",
      create(s"$cache-4").mode(PARTITIONED).backups(1) as "4",
      create(s"$cache-5").backups(1).atomicity(ATOMIC).mode(REPLICATED) as "5"
    )
  setUp(scn.inject(atOnceUsers(1))).protocols(protocol).assertions(global.failedRequests.count.is(0))
}

/**
 * Create cache using full ClientCacheConfiguration instance.
 */
class CreateCacheThinConfigSimulation extends Simulation with IgniteSupport with StrictLogging {
  private val cache = "cache-thin"
  private val scn = scenario("Basic")
    .feed(feeder)
    .ignite(create(cache).cfg(new ClientCacheConfiguration()))
  setUp(scn.inject(atOnceUsers(1))).protocols(protocol).assertions(global.failedRequests.count.is(0))
}

/**
 * Create cache using full CacheConfiguration instance.
 */
class CreateCacheThickConfigSimulation extends Simulation with IgniteSupport with StrictLogging {
  private val cache = "cache-thick"
  private val scn = scenario("Basic")
    .feed(feeder)
    .ignite(create(cache).cfg(new CacheConfiguration[Int, Int]()) as "create")
  setUp(scn.inject(atOnceUsers(1))).protocols(protocol).assertions(global.failedRequests.count.is(0))
}
